src/impl/tpls/KokkosSparse_spmv_mv_tpl_spec_decl.hpp

```diff
@@ -0,0 +1,414 @@
+//       Copyright (2020) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Siva Rajamanickam (srajama@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOSPARSE_SPMV_MV_TPL_SPEC_DECL_HPP_
+#define KOKKOSPARSE_SPMV_MV_TPL_SPEC_DECL_HPP_
+
+#include "KokkosKernels_Controls.hpp"
+
+#ifdef KOKKOSKERNELS_ENABLE_TPL_CUSPARSE
+
+/* CUSPARSE_VERSION < 10301 either doesn't have cusparseSpMM
+   or the non-tranpose version produces incorrect results.
+*/
+#if defined(CUSPARSE_VERSION) && (10301 <= CUSPARSE_VERSION)
+#include "cusparse.h"
+#include "KokkosSparse_Utils_cusparse.hpp"
+
+namespace KokkosSparse {
+namespace Impl {
+
+template <typename T>
+cudaDataType cuda_data_type_from(const std::string &msg) {
+  throw std::logic_error(msg);
+}
+
+#if defined(KOKKOS_HALF_T_IS_FLOAT) && !KOKKOS_HALF_T_IS_FLOAT
+template <>
+inline cudaDataType cuda_data_type_from<Kokkos::Experimental::half_t>(
+    const std::string &) {
+  return CUDA_R_16F;  // Kokkos half_t is a half
+}
+#else
+template <>
+inline cudaDataType cuda_data_type_from<Kokkos::Experimental::half_t>(
+    const std::string &) {
+  return CUDA_R_32F;  // Kokkos half_t is a float
+}
+#endif
+template <>
+inline cudaDataType cuda_data_type_from<float>(const std::string &) {
```

Comment:
Very similar logic already exists in `src/impl/tpls/KokkosSparse_spmv_tpl_spec_decl.hpp`, we should consolidate it in `KokkosSparse_Utils_cusparse.hpp` to avoid re-implementing it for every kernel.

Reply:
I think we can consolidate the direct translation of c++ types to cuSparse types for the operands, however the "compute type" (i.e. the precision at which the computation is done) is not necessarily the same as or even selected from one of the operands, so that should stay specific to each kernel.